### PR TITLE
[RISCV] Generate RISCVISAInfo table from RISCVFeatures.td.

### PR DESCRIPTION
**Pull request by topperc**
This generates the SupportedExtensions and ImpliedExts information from the RISCVExtension records found in RISCVFeatures.td.

Some of the extensions listed in the individual `ImpliedExts*` arrays may be in a different, but the order in those array doesn't matter. I manually verified the all the extensions were still present in each array.

I've added the new information to the existing RISCVTargetParserDef.inc and RISCVTargetDefEmitter.cpp so we don't need to re-parse the entirety of RISCV.td a second time for a new file.

### CODE REVIEW

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 1]
```diff
@@ -47,204 +47,8 @@ static const char *RISCVGImplications[] = {
...
-    {"zicfiss", {0, 4}},
-
-    {"ztso", {0, 1}},
-
-    {"zvfbfmin", {1, 0}},
-    {"zvfbfwma", {1, 0}},
-};
-// clang-format on
+#define GET_SUPPORTED_EXTENSIONS
+#include "llvm/TargetParser//RISCVTargetParserDef.inc"
```
Comment (dtcxzyw) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:51]:
```suggestion
#include "llvm/TargetParser/RISCVTargetParserDef.inc"
```

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 2]
```diff
@@ -1112,67 +856,8 @@ struct ImpliedExtsEntry {
   bool operator<(StringRef Other) const { return Name < Other; }
 };
 
-// Note: The table needs to be sorted by name.
```
Comment (michaelmaitland) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:1115]: IIRC there may have been some things that depended on the order here, hence the comment. But the PR description says the order is different. Does that break anything?

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 3]
```diff
@@ -47,204 +47,8 @@ static const char *RISCVGImplications[] = {
...
-    {"zvl4096b", {1, 0}},
-    {"zvl512b", {1, 0}},
-    {"zvl64b", {1, 0}},
-    {"zvl65536b", {1, 0}},
-    {"zvl8192b", {1, 0}},
-};
-
-// NOTE: This table should be sorted alphabetically by extension name.
-// clang-format off
-static const RISCVSupportedExtension SupportedExperimentalExtensions[] = {
```
Comment (michaelmaitland) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:224]: It looks like this patch removes building of `SupportedExperimentalExtensions`. Was it unused before this patch?

File: llvm/lib/TargetParser/RISCVISAInfo.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:1115]: The patch changed the order of the strings in the `ImpliedExts*` arrays above. But the order in those arrays don't matter.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp (hunk 3 above)
Comment (topperc) [reply llvm/lib/TargetParser/RISCVISAInfo.cpp:224]: This patch does not remove it. See this code. We call `printExtensionTable` twice.

```
static void printExtensionTable(raw_ostream &OS,
                                const std::vector<Record *> &Extensions,
                                bool Experimental) {
  OS << "static const RISCVSupportedExtension Supported";
  if (Experimental)
    OS << "Experimental";
  OS << "Extensions[] = {\n";
```

**Review by 4vtomat — APPROVED** [review]
LGTM~

**Review by wangpc-pp — COMMENTED** [review]
Add tests to `llvm/test/TableGen/riscv-target-def.td`?

File: llvm/utils/TableGen/RISCVTargetDefEmitter.cpp [hunk 4]
```diff
@@ -101,5 +197,10 @@ static void EmitRISCVTargetDef(RecordKeeper &RK, raw_ostream &OS) {
   OS << "\n#undef TUNE_PROC\n";
 }
 
+static void EmitRISCVTargetDef(RecordKeeper &RK, raw_ostream &OS) {
+  emitRISCVExtensions(RK, OS);
+  emitRISCVProcs(RK, OS);
+}
+
 static TableGen::Emitter::Opt X("gen-riscv-target-def", EmitRISCVTargetDef,
                                 "Generate the list of CPU for RISCV");
```
Comment (wangpc-pp) [inline llvm/utils/TableGen/RISCVTargetDefEmitter.cpp:206]: This help info is not correct now, and maybe there are more places (like comments in line 9-10) to reword.

File: llvm/test/TableGen/riscv-target-def.td [hunk 5]
```diff
@@ -75,22 +81,48 @@ def ROCKET_RV64 : RISCVProcessorModel<"rocket-rv64",
...
+// CHECK-NEXT: #undef GET_SUPPORTED_EXTENSIONS
+
+// CHECK:      static const RISCVSupportedExtension SupportedExtensions[] = {
+// CHECK-NEXT:     {"f", {2, 2}},
+// CHECK-NEXT:     {"i", {2, 1}},
+// CHECK-NEXT:     {"zicsr", {2, 0}},
+// CHECK-NEXT:     {"zifencei", {2, 0}},
+// CHECK-NEXT: };
+
+// CHECK:      static const RISCVSupportedExtension SupportedExperimentalExtensions[] = {
```
Comment (wangpc-pp) [inline llvm/test/TableGen/riscv-target-def.td:94]: Can we add a dummy experimental extension?

File: llvm/utils/TableGen/RISCVTargetDefEmitter.cpp [hunk 6]
```diff
@@ -17,6 +17,102 @@
...
+  return Name;
+}
+
+static void emitRISCVExtensions(RecordKeeper &Records, raw_ostream &OS) {
+  OS << "#ifdef GET_SUPPORTED_EXTENSIONS\n";
+  OS << "#undef GET_SUPPORTED_EXTENSIONS\n\n";
+
+  std::vector<Record *> Extensions =
+      Records.getAllDerivedDefinitions("RISCVExtension");
+  llvm::sort(Extensions, LessRecordExtensionName());
```
Comment (wangpc-pp) [inline llvm/utils/TableGen/RISCVTargetDefEmitter.cpp:72]: Use lambda?

**Review by wangpc-pp — APPROVED** [review]
LGTM. Thanks!

**Review by dtcxzyw — APPROVED** [review]

**Pull request merged**